### PR TITLE
Test recipe

### PR DESCRIPTION
**Pull request by dylan-muszel**
### Summary
This is a recipe where the only importants are the test. I also add a `WolmoActivityTest` and a `WolmoPresenterTest` that in my opinion should be in wolmo.
Also, I migrate it to AndroidX.

### CODE REVIEW

File: app/build.gradle [hunk 1]
```diff
@@ -152,16 +152,15 @@ dependencies {
 
     // Testing dependencies
```
Comment (emalamela) [inline app/build.gradle]: Just to be organized, I believe we should put all of the `Testing Dependencies` at the bottom. What do you think?

File: app/src/main/java/ar/com/wolox/android/cookbook/tests/TestLoginRecipeFragment.kt [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+import ar.com.wolox.android.cookbook.R
+import ar.com.wolox.android.cookbook.datasync.DataSyncRecipeActivity
+import ar.com.wolox.wolmo.core.fragment.WolmoFragment
+import ar.com.wolox.wolmo.core.util.ToastFactory
+import kotlinx.android.synthetic.main.fragment_test_login.vTestLoginEmailInput
+import kotlinx.android.synthetic.main.fragment_test_login.vTestLoginLoginBtn
+import kotlinx.android.synthetic.main.fragment_test_login.vTestLoginPasswordInput
+import javax.inject.Inject
+
+class TestLoginRecipeFragment : WolmoFragment<TestLoginRecipePresenter>(), TestLoginRecipeView {
```
Comment (emalamela) [inline app/src/main/java/ar/com/wolox/android/cookbook/tests/TestLoginRecipeFragment.kt]: This `Fragment` is not very `MVP`-esque. All the validations and business logic decisions are on the `Fragment` side.

File: app/src/androidTest/java/ar/com/wolox/android/cookbook/tests/WolmoActivityTest.kt [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+@RunWith(AndroidJUnit4::class)
+open class WolmoActivityTest<T : WolmoActivity> {
+
+    @Rule
+    @JvmField
+    var activityRule = ActivityTestRule<T>(getActivityClass())
+
+    @Before
+    fun beforeWolmoActivity() {
```
Comment (emalamela) [inline app/src/androidTest/java/ar/com/wolox/android/cookbook/tests/WolmoActivityTest.kt]: I'm not sure I understand what this `@Before` actually does.

File: app/src/main/java/ar/com/wolox/android/cookbook/tests/TestLoginRecipeService.kt [hunk 4]
```diff
@@ -0,0 +1,17 @@
+package ar.com.wolox.android.cookbook.tests
+
+import ar.com.wolox.wolmo.core.di.scopes.ApplicationScope
+import javax.inject.Inject
+
+@ApplicationScope
+open class TestLoginRecipeService @Inject constructor() {
```
Comment (emalamela) [inline app/src/main/java/ar/com/wolox/android/cookbook/tests/TestLoginRecipeService.kt]: This should perform an async operation to better emulate the behavior of an API, for example.

File: app/src/main/res/values/strings.xml [hunk 5]
```diff
@@ -40,4 +41,12 @@
     <string name="facebook_login_error_unexpected">Unexpected error</string>
     <string name="facebook_login_btn">Login with Facebook</string>
 
+    <!-- Test Login -->
+    <string name="test_login_email">Email</string>
+    <string name="test_login_password">Password</string>
+	<string name="test_login_login">Login</string>
```
Comment (emalamela) [inline app/src/main/res/values/strings.xml]: Indentation.

File: app/src/main/res/values/strings.xml [hunk 6]
```diff
@@ -40,4 +41,12 @@
...
+    <string name="test_login_empty_field">Debe completar este campo</string>
+    <string name="test_login_invalid_email">Email invalido</string>
+	<string name="test_login_login_error">Incorrect user or password.</string>
```
Comment (emalamela) [inline app/src/main/res/values/strings.xml]: Indentation.

File: app/src/androidTest/java/ar/com/wolox/android/cookbook/tests/TestLoginRecipeActivityTest.kt [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+import androidx.test.espresso.Espresso.onView
+import androidx.test.espresso.action.ViewActions.click
+import androidx.test.espresso.action.ViewActions.typeText
+import androidx.test.espresso.assertion.ViewAssertions.matches
+import androidx.test.espresso.matcher.ViewMatchers.withId
+import ar.com.wolox.android.cookbook.R
+import org.junit.Test
+
+class TestLoginRecipeActivityTest : WolmoActivityTest<TestLoginRecipeActivity>() {
```
Comment (emalamela) [inline app/src/androidTest/java/ar/com/wolox/android/cookbook/tests/TestLoginRecipeActivityTest.kt]: I believe we should test this over the `Fragment`, not the `Activity`.

File: app/build.gradle (hunk 1 above)
Comment (dylan-muszel) [reply app/build.gradle]: 👍

Comment (emalamela) [reply app/build.gradle]: I like it!

File: app/src/main/AndroidManifest.xml [hunk 8]
```diff
@@ -1,76 +1,66 @@
-<manifest
+<?xml version="1.0" encoding="utf-8"?>
+<manifest package="ar.com.wolox.android.cookbook"
```
Comment (emalamela) [inline app/src/main/AndroidManifest.xml]: Namespace declaration below.

File: app/src/main/java/ar/com/wolox/android/cookbook/tests/TestLoginRecipePresenter.kt [hunk 9]
```diff
@@ -6,6 +6,34 @@ import javax.inject.Inject
...
-        loginService.login(email, password)?.let { view.goToNextWindow() } ?: view.showLoginError()
+        val isValidEmail = verifyEmail(email)
+        val isValidPassword = verifyPassword(password)
+
+        if (isValidEmail && isValidPassword) {
+            loginService.login(email, password, { view.goToNextWindow() }, { view.showLoginError() })
+        }
+    }
+
+    private fun verifyEmail(email: String): Boolean {
```
Comment (emalamela) [inline app/src/main/java/ar/com/wolox/android/cookbook/tests/TestLoginRecipePresenter.kt]: This function should only return whether the `email` is valid and the caller show whatever it has to show. It's doing too much.

File: app/src/main/java/ar/com/wolox/android/cookbook/tests/TestLoginRecipePresenter.kt [hunk 10]
```diff
@@ -6,6 +6,34 @@ import javax.inject.Inject
...
+            view.showInvalidEmailError()
+            return false
+        }
+
+        return true
+    }
+
+    private fun isValidEmail(email: String) = android.util.Patterns.EMAIL_ADDRESS.matcher(email).matches()
+
+    private fun verifyPassword(password: String): Boolean {
```
Comment (emalamela) [inline app/src/main/java/ar/com/wolox/android/cookbook/tests/TestLoginRecipePresenter.kt]: Same as `verifyEmail`.

**Pull request merged**